L1Trigger/TrackFindingTracklet/src/PurgeDuplicate.cc

```diff
@@ -323,26 +323,29 @@ void PurgeDuplicate::execute(std::vector<Track>& outputtracks_, unsigned int iSe
 
     // Make the final track objects, fit with KF, and send to output
     for (unsigned int itrk = 0; itrk < numStublists; itrk++) {
-      Tracklet* tracklet = inputtracklets_[itrk];
-      std::vector<const Stub*> trackstublist = inputstublists_[itrk];
-
-      HybridFit hybridFitter(iSector, settings_, globals_);
-      hybridFitter.Fit(tracklet, trackstublist);
-
-      // If the track was accepted (and thus fit), add to output
-      if (tracklet->fit()) {
-        // Add track to output if it wasn't merged into another
-        Track* outtrack = tracklet->getTrack();
-        outtrack->setSector(iSector);
-        if (trackInfo[itrk].second == true)
-          outtrack->setDuplicate(true);
-        else
-          outputtracklets_[trackInfo[itrk].first]->addTrack(tracklet);
-
-        // Add all tracks to standalone root file output
-        outtrack->setStubIDpremerge(inputstubidslists_[itrk]);
-        outtrack->setStubIDprefit(mergedstubidslists_[itrk]);
-        outputtracks_.push_back(*outtrack);
+      bool duplicateTrack = trackInfo[itrk].second;
+      if (not duplicateTrack) {
+        Tracklet* tracklet = inputtracklets_[itrk];
+        std::vector<const Stub*> trackstublist = inputstublists_[itrk];
+
+        HybridFit hybridFitter(iSector, settings_, globals_);
+        hybridFitter.Fit(tracklet, trackstublist);
+
+        // If the track was accepted (and thus fit), add to output
+        if (tracklet->fit()) {
+          // Add track to output if it wasn't merged into another
+          Track* outtrack = tracklet->getTrack();
+          outtrack->setSector(iSector);
+          if (trackInfo[itrk].second == true)
+            outtrack->setDuplicate(true);
+          else
+            outputtracklets_[trackInfo[itrk].first]->addTrack(tracklet);
+
+          // Add all tracks to standalone root file output
+          outtrack->setStubIDpremerge(inputstubidslists_[itrk]);
+          outtrack->setStubIDprefit(mergedstubidslists_[itrk]);
+          outputtracks_.push_back(*outtrack);
```

Comment:
Duplicate tracks are now excluded from `outputtracks_`, but only in the case of the "merge" removal type. For other removal types, all tracks are included:
https://github.com/cms-L1TK/cmssw/blob/7ac18f0153eb641531b11bc21c2ab196dfa93f65/L1Trigger/TrackFindingTracklet/src/PurgeDuplicate.cc#L475-L476
This doesn't affect the behavior of L1FPGATrackProducer because duplicate tracks are explicitly skipped here:
https://github.com/cms-L1TK/cmssw/blob/7ac18f0153eb641531b11bc21c2ab196dfa93f65/L1Trigger/TrackFindingTracklet/plugins/L1FPGATrackProducer.cc#L574-L575
But should `outputtracks_` be made consistent across removal types?

Reply:
@skinnari believes the reason PurgeDuplicates doesn't delete duplicates from the Track collection, but merely flags them as duplicates, is to make it easier to study the PD performance. There is therefore justification in keeping this unchanged for the pure Tracklet algo. However, for Hybrid, since PurgeDuplicates doesn't simply run the duplicate removal but also the KF, it would be harder to use this feature. In addition, the duplicate tracks after PurgeDuplicates could not be kept without wasting CPU by running the KF on all of them.

Reply:
OK I think that argument makes sense.